rbd: add kmip encryption type

Rakshith-R opened the pull request:
The Key Management Interoperability Protocol (KMIP)
 is an extensible communication protocol
that defines message formats for the manipulation
of cryptographic keys on a key management server.
Ceph-CSI can now be configured to connect to
various KMS using KMIP for encrypting RBD volumes.

https://en.wikipedia.org/wiki/Key_Management_Interoperability_Protocol

Resolves: https://github.com/ceph/ceph-csi/issues/3282

Signed-off-by: Rakshith R <rar@redhat.com>

Heavily inspired from https://github.com/noobaa/noobaa-operator/pull/964#issue-1311434745

humblec commented on docs/deploy-rbd.md:
```diff
@@ -415,6 +415,33 @@ the AWS KMS is expected to contain:
+#### Configuring KMIP KMS
+
+The Key Management Interoperability Protocol (KMIP) is an extensible communication protocol that defines message formats for the manipulation of cryptographic keys on a key management server.
+Ceph-CSI can be configured to connect to various KMS servers using 
```
KMS/KMIP ?

humblec commented on examples/kms/vault/kmip-credentials.yaml:
```diff
@@ -0,0 +1,13 @@
+apiVersion: v1
+kind: Secret
+metadata:
+  name: ceph-csi-kp-credentials
```
ceph-csi-kmip-credentials?

humblec commented:
@Rakshith-R looks like `ans` field has to be renamed for the codespell to be happy :)

```./internal/kms/kms_util.go:37: ans ==> and
./internal/kms/kms_util.go:49: ans ==> and
./internal/kms/kms_util.go:61: ans ==> and
./internal/kms/kms_util.go:71: ans ==> and
./internal/kms/kms_util.go:81: ans ==> and
```

nixpanic commented:
Please split the vendor part from the actual feature, that makes it much easier to review.

Rakshith-R committed:
rbd: add kmip encryption type

The Key Management Interoperability Protocol (KMIP)
is an extensible communication protocol
that defines message formats for the manipulation
of cryptographic keys on a key management server.
Ceph-CSI can now be configured to connect to
various KMS using KMIP for encrypting RBD volumes.

https://en.wikipedia.org/wiki/Key_Management_Interoperability_Protocol

Signed-off-by: Rakshith R <rar@redhat.com>

Rakshith-R committed:
rebase: vendor files required for kmip

Signed-off-by: Rakshith R <rar@redhat.com>

SanjalKatiyar commented on docs/deploy-rbd.md:
```diff
@@ -415,6 +415,33 @@ the AWS KMS is expected to contain:
+1. `CA_CERT`: CA certificate that will be used to connect to KMIP server.
+2. `CLIENT_CERT`: Client certificate that will be used to connect to KMIP server.
+3. `CLIENT_KEY`: Client key that will be used to connect to KMIP server
+4. `UNIQUE_IDENTIFIER`: Unique ID of the key to use for encrypting/decrypting.
```
IIUC `UNIQUE_IDENTIFIER` is for encrypting/decrypting DEK.
But, do we want this identifier as an input from the UI ? Can't we do the same as we are doing for HPCS ?

Rakshith-R replied:
> IIUC `UNIQUE_IDENTIFIER` is for encrypting/decrypting DEK.
> But, do we want this identifier as an input from the UI ? Can't we do the same as we are doing for HPCS ?

@SanjalKatiyar 
Yes this needs to be an created in the secret, 
We are doing it exactly as in hpcs.
This field is similar to 
https://github.com/ceph/ceph-csi/blob/2fc10ded656c4f587b6a9763f4af3b5147b28b01/examples/kms/vault/kp-credentials.yaml#L11

SanjalKatiyar replied:
If I am correct, customer_root_key was something that was supported/required by HPCS KMS instance itself...
but in our case we will not be passing `UNIQUE_IDENTIFIER` to the KMS instance, so we are introducing new parameter which customers do not know about (I mean not supported by Cipher Trust Manager itself)...
Just wondering if it is okay to do so (we can discuss this in tomorrow meeting once as well !!) ??

Rakshith-R replied:
marking it resolved based on discussion

SanjalKatiyar commented on examples/kms/vault/csi-kms-connection-details.yaml:
```diff
@@ -74,5 +74,14 @@ data:
+    {
+      "KMS_PROVIDER": "kmip",
+      "KMIP_ENDPOINT": "kmip:5696",
+      "KMIP_SECRET_NAME": "ceph-csi-kmip-credentials",
```
**Just fyi**, since we will be creating this Secret from UI, so will add some random generated text to the Secret's name and pass that name to this "KMIP_SECRET_NAME" key.
Reason for adding random text to Secret's name is that, user can create multiple KMS connections, if 2 different connections will have same name for Secret it could be an issue.

SanjalKatiyar replied:
also, Secret will be in deployment namespace, `openshift-storage` in our case...

Rakshith-R replied:
Yes, and yes
Both sounds good to me.

humblec commented on internal/kms/kmip.go:
```diff
@@ -0,0 +1,512 @@
+	kmsTypeKMIP = "kmip"
+
+	// KMIPDefaulReadTimeout is the default read network timeout.
+	KMIPDefaulReadTimeout = 10
```
KMIPDefaultReadOut. also in next variable @Rakshith-R

Rakshith-R replied:
@humblec I did'nt get you

Madhu-1 replied:
you have a typo `Defaul` to `Default`

humblec commented on internal/kms/kmip.go:
```diff
@@ -0,0 +1,512 @@
+	return kms, nil
+}
+
+// EncryptDEK uses the Amazon KMS and the configured CMK to encrypt the DEK.
```
@Rakshith-R remove above Amazon reference.

humblec commented on internal/kms/kmip.go:
```diff
@@ -0,0 +1,512 @@
+	return string(emdData), nil
+}
+
+// DecryptDEK uses the Amazon KMS and the configured CMK to decrypt the DEK.
```
same here

humblec commented on internal/kms/kms_util.go:
```diff
@@ -0,0 +1,43 @@
+import "fmt"
+
+// setConfigInt fetches a value from a configuration map and converts it to
+// a integer.
```
s/a/an

Madhu-1 commented on internal/kms/kmip.go:
```diff
@@ -0,0 +1,512 @@
+	KMIPDefaulReadTimeout = 10
+
+	// KMIPDefaulWriteTimeout is the default write network timeout.
+	KMIPDefaulWriteTimeout = 10
```
no need to export these variables?

Madhu-1 commented on internal/kms/kmip.go:
```diff
@@ -0,0 +1,512 @@
+	endpoint         string
+	tlsConfig        *tls.Config
+	uniqueIdentifier string
+	readTimeout      int
```
use int8 instead of int?

Madhu-1 commented on internal/kms/kmip.go:
```diff
@@ -0,0 +1,512 @@
+		return nil, fmt.Errorf("failed to get secrets: %w", err)
+	}
+
+	var found bool
```
why this var is required? is it due to a linter error?

Rakshith-R replied:
removed this

Madhu-1 commented on internal/kms/kmip.go:
```diff
@@ -0,0 +1,512 @@
+	}
+
+	var found bool
+	caCert, found := secrets[kmipCACert]
+	if !found {
+		return nil, fmt.Errorf("%w: %s", errConfigOptionMissing, kmipCACert)
+	}
+
+	clientCert, found := secrets[kmipCLientCert]
+	if !found {
+		return nil, fmt.Errorf("%w: %s", errConfigOptionMissing, kmipCLientCert)
+	}
+
+	clientKey, found := secrets[kmipClientKey]
+	if !found {
+		return nil, fmt.Errorf("%w: %s", errConfigOptionMissing, kmipCLientCert)
+	}
+
+	kms.uniqueIdentifier, found = secrets[kmipUniqueIdentifier]
+	if !found {
+		return nil, fmt.Errorf("%w: %s", errConfigOptionMissing, kmipUniqueIdentifier)
+	}
```
these are optional or must have?

Rakshith-R replied:
> these are optional or must have?

These are must haves

Madhu-1 commented on internal/kms/kmip.go:
```diff
@@ -0,0 +1,512 @@
+		return "", err
+	}
+
+	bi, err := kms.response(respMsg, kmip14.OperationEncrypt, uniqueBatchItemID)
```
we dont need to close any response body, right? just seeing if is it going to cause any memory leak.

Rakshith-R replied:
There is no need for that i guess.

Madhu-1 commented on internal/kms/kmip.go:
```diff
@@ -0,0 +1,512 @@
+		return "", err
+	}
+
+	bi, err := kms.response(respMsg, kmip14.OperationDecrypt, uniqueBatchItemID)
```
what is `bi`? use a better name with more readability?

Rakshith-R replied:
renaming to batchItem

Madhu-1 commented on internal/kms/kmip.go:
```diff
@@ -0,0 +1,512 @@
+	if err != nil {
+		return nil, err
+	}
+	if kms.readTimeout != 0 {
```
if an error close the connection in defer

Madhu-1 commented on internal/kms/kmip.go:
```diff
@@ -0,0 +1,512 @@
+}
+
+// Verify the response success and return the batch item.
+func (kms *kmipKMS) response(
```
rename the function name to verifyResponse to match the work its doing?

Madhu-1 commented on internal/kms/kmip.go:
```diff
@@ -0,0 +1,512 @@
+	return &bi, nil
+}
+
+// TODO: use the following structs from https://github.com/gemalto/kmip-go
```
is there any Issue opened in https://github.com/gemalto/kmip-go for tracking?

Rakshith-R replied:
https://github.com/ThalesGroup/kmip-go/issues/21

humblec commented:
@Rakshith-R can you please address the comments.. I am planning to start the Release work of 3.7, thats why :)

nixpanic commented on internal/kms/kmip.go:
```diff
@@ -0,0 +1,514 @@
+	return conn, nil
+}
+
+// discover performs KMIP handshake.
```
It would be useful to link to the documentation (+chapter/paragraph) where this handshake is explained. Maybe for other protocol procedures too.

Rakshith-R replied:
Its just discover protocol being run, linked discover operation chapter here.

nixpanic commented on internal/kms/kmip.go:
```diff
@@ -0,0 +1,518 @@
+	}
+
+	// KMIP handshake
+	err = kms.discover(conn)
```
just above is `conn.Handshake()`... what's the difference?

Rakshith-R replied:
removed the term handshake, added more explanation in function description.

nixpanic commented on internal/kms/kmip.go:
```diff
@@ -0,0 +1,518 @@
+func (kms *kmipKMS) connect() (*tls.Conn, error) {
+	conn, err := tls.Dial("tcp", kms.endpoint, kms.tlsConfig)
+	if err != nil {
+		return nil, err
```
ideally add some more context in the returned errors with `fmt.Errorf(...)`

nixpanic commented on internal/kms/kmip.go:
```diff
@@ -0,0 +1,518 @@
+			},
+		})
+	if err != nil {
+		return err
```
make sure that returned errors are unique, so that tracking the source of the error trivial. Use `fmt.Errorf(...)` for returned errors if you can.

nixpanic commented on internal/kms/kmip.go:
```diff
@@ -0,0 +1,518 @@
+
+	req, err := ttlv.Marshal(msg)
+	if err != nil {
+		return nil, nil, nil, err
```
same here, use `fmt.Errorf()` so that tracking the source of errors becomes easy

nixpanic commented on internal/kms/kmip.go:
```diff
@@ -0,0 +1,518 @@
+	return nil
+}
+
+// send sends KMIP operation over tls connection, return response and error.
```
return response and error... but there are 4 return values? probably good to explain each of them

nixpanic reviewed (changes requested):
mostly nits

Add a note somewhere about how it was tested?

Rakshith-R commented:
> mostly nits
> 
> Add a note somewhere about how it was tested?

This was tested with pykimp server instance deployed in kubernetes cluster, 
I'll open an issue to track this, so we can add e2e for kmip encryption in a follow-up pr.

Rakshith-R commented:
addressed the comments PTAL @nixpanic @humblec

nixpanic commented on internal/kms/kmip.go:
```diff
@@ -0,0 +1,527 @@
+	kmip "github.com/gemalto/kmip-go"
+	"github.com/gemalto/kmip-go/kmip14"
+	"github.com/gemalto/kmip-go/ttlv"
+	"github.com/google/uuid"
```
we use github.com/pborman/uuid already, maybe create a new issue to move to github.com/google/uuid instead? The google one is used by IBM Keyprotect and Kubernetes packages already, pborman only in `internal/journal/voljournal.go`

Rakshith-R replied:
https://github.com/ceph/ceph-csi/issues/3315

nixpanic replied:
ok, thanks!

nixpanic commented on internal/kms/kmip.go:
```diff
@@ -0,0 +1,527 @@
+		return nil, fmt.Errorf("%w: %s", errConfigOptionMissing, kmipCACert)
+	}
+
+	clientCert, found := secrets[kmipCLientCert]
```
As this is required, a client can only connect to the KMIP server with its own client certificate? That is not very common for TLS connections. Where does the KMIP standard describe this requirement?

Rakshith-R replied:
https://thalesdocs.com/ctp/cm/latest/reference/kmip-ref/

There are a couple of options of authenticating with kmip server, currently we are going this route. 
We can adapt this based later as required, since this just preliminary exploration with kmip encryption type.

nixpanic replied:
OK, so this is a common process for Thales CyberTrust. Have you considered adding an option like `"AUTH_METHOD": "TLSCERT"` or something? That suggests making it more modular in the future. Of course, this can be the default if no `AUTH_METHOD` is set too...

Rakshith-R replied:
https://github.com/ceph/ceph-csi/issues/3316
noted :+1:

nixpanic commented on internal/kms/kmip.go:
```diff
@@ -0,0 +1,527 @@
+		return nil, fmt.Errorf("%w: %s", errConfigOptionMissing, kmipCLientCert)
+	}
+
+	clientKey, found := secrets[kmipClientKey]
```
the clientKey is used to decrypt the clientCert, what if the clientCert is not encrypted?

Rakshith-R replied:
> the clientKey is used to decrypt the clientCert, what if the clientCert is not encrypted?

This is the scenario that is currently tested/supported, 
we can maybe add support to such scenario as and when required after testing.

nixpanic replied:
sure, that's fine. I hope client certificates are encrypted in the common scenarios anyway.
